exasol/ai/mcp/server/main.py

```diff
@@ -328,56 +335,58 @@ def get_connection_factory(
-      above conditions cannot be met. For this option we will not cache the connection,
-      otherwise a token issued for one user could be used to execute a query requested
-      by another user. As in the first option, the permissions of the server's DB user
-      must be set accordingly.
+         token with the same identity provider as the MCP server. The subject, the DB
+         user is identified with in the database, should, according to RFC 9068, match
+         the subject field in the access token issued to this user.
+
+    - The last option is a blend between the first two. It works in a multiuser mode,
+      when the first of the above conditions is met but the second is not. The connection
+      is opened using the pre-configured database credentials, as in the first option.
+      But since the actual username can be identified, the connection impersonates this
+      user. All subsequent queries are executed under this user's permissions. For this
+      to work the server's user must have the "IMPERSONATE ANY USER" or "IMPERSONATION ON
+      <user/role>" privilege.
     """
-    common_kwargs, use_open_id = _create_connection_kwargs(env, **extra_kwargs)
+    common_kwargs = _create_connection_kwargs(env, **extra_kwargs)
     connection_pool = _create_connection_pool(env)
 
     @contextmanager
     def connection_factory() -> Generator[ExaConnection, None, None]:
-        # In the OpenID mode try to get the current username, as well as the access
-        # token, from the MCP context. That failed, use the default username.
-        # Also, use the default username in all other modes.
-        use_pool = True
-        if use_open_id:
-            user, token = get_oidc_user(env.get(ENV_USERNAME_CLAIM))
-            if not user:
-                if ENV_USER not in env:
-                    raise RuntimeError(
-                        "Cannot extract database username from the MCP "
-                        "context, and no default username is specified"
-                    )
-                user = env[ENV_USER]
-                use_pool = False
-        else:
-            user, token = env[ENV_USER], ""
-
-        # Try to get the connection for the current user from the pool,
-        # unless the pool is disabled.
-        connection = connection_pool.checkout(user) if use_pool else None
+        # Try to get the actual username and the access token from the MCP context.
+        oidc_user, token = get_oidc_user(env.get(ENV_USERNAME_CLAIM))
+        server_user = env.get(ENV_USER)
+        user = oidc_user or server_user
+        if not user:
+            raise RuntimeError(
+                "Cannot extract database username from the MCP context, "
+                "and default username is not specified."
+            )
+
+        # Try to get the connection for the current user from the pool.
+        connection = connection_pool.checkout(user)
```

Comment:
If both (oidc_user and server_user) are set, here you prefer the `oidc_user`, but below line 374 you prefer the `server_user`.

Reply:
If you say 
`# Always prefer to connect with pre-configured server credentials.`, shouldn't this also count for the connection from the connection pool?

Reply:
There are two question regarding the name.
1. What name do we use when we actually make the connection?
2. How we remember this connection, i.e. what name does it go to the pool with?

The answers are:
1. If the server's credentials are provided (`server_user`) we use that in preference of the discovered user name (`oidc_user`)
2. This is the name of the user under which identity the queries are executed. We can open the connection with `server_user`, but then impersonate `oidc_user`. In that case the connection is remembered under the name of `oidc_user`.

Reply:
Got it